features/mbedtls/targets/TARGET_Silicon_Labs/crypto_aes.c

```diff
@@ -0,0 +1,545 @@
+/*
+ *  FIPS-197 compliant AES implementation
+ *
+ *  Copyright (C) 2017, Silicon Labs, http://www.silabs.com
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * This file includes alternative plugin implementations of various
+ * functions in aes.c using the CRYPTO hardware accelerator incorporated
+ * in MCU devices from Silicon Laboratories.
+ */
+
+/*
+ *  The AES block cipher was designed by Vincent Rijmen and Joan Daemen.
+ *
+ *  http://csrc.nist.gov/encryption/aes/rijndael/Rijndael.pdf
+ *  http://csrc.nist.gov/publications/fips/fips197/fips-197.pdf
+ */
+
+#include "mbedtls/aes.h"
+
+#if defined(MBEDTLS_AES_C)
+#if defined(MBEDTLS_AES_ALT)
+
+#include "em_device.h"
+
+#if defined(CRYPTO_PRESENT)
+
+#include "crypto_management.h"
+#include "em_crypto.h"
+#include "em_core.h"
+#include <string.h>
+
+__STATIC_INLINE void CRYPTO_DataReadUnaligned(volatile uint32_t * reg,
+                                               const uint8_t * val)
+{
+  /* Check data is 32bit aligned, if not, read into temporary buffer and
+     then move to user buffer. */
+  if ((uint32_t)val & 0x3)
+  {
+    uint32_t temp[4];
+    CRYPTO_DataRead(reg, temp);
+    memcpy((void*)val, temp, 16);
+  }
+  else
+  {
+    CRYPTO_DataRead(reg, (uint32_t*)val);
+  }
+}
+
+__STATIC_INLINE void CRYPTO_DataWriteUnaligned(volatile uint32_t * reg,
+                                               const uint8_t * val)
+{
+  /* Check data is 32bit aligned, if not move to temporary buffer before
+     writing.*/
+  if ((uint32_t)val & 0x3)
+  {
+    uint32_t temp[4];
```

Comment:
The only case I know of where a uint32_t would potentially be stored out of its native alignment is in a packed struct (which is not the case here)

Reply:
As I mentioned in my previous comment, I do not think there are any guarantees that anything is stored in the desired alignment. The compiler can potentially do whatever it wants. I suggest that we simply use the `MBED_ALIGN()` macro that is exactly for this purposes:
https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_toolchain.h#L56

Reply:
MBED_ALIGN is not a solution since we're shipping the same codebase to a broader audience, and mbed_toolchain.h is not a part of the mbed TLS codebase.
Please keep in mind that these files are drivers for mbed TLS, not mbed OS.